src/libs/print_settings.c

```diff
@@ -280,101 +271,187 @@ _print_button_clicked (GtkWidget *widget, gpointer user_data)
-  free (pdf_page);
+  if(dt_control_job_get_state(job) == DT_JOB_STATE_CANCELLED) return 0;
+  dt_control_job_set_progress(job, 0.95);
 
   // send to CUPS
 
-  dt_print_file (imgid, filename, &ps->prt);
-
-  g_unlink(filename);
+  dt_print_file (params->imgid, params->pdf_filename, params->job_title, &params->prt);
+  dt_control_job_set_progress(job, 1.0);
 
   // add tag for this image
 
   char tag[256] = { 0 };
   guint tagid = 0;
-  snprintf (tag, sizeof(tag), "darktable|printed|%s", ps->prt.printer.name);
+  snprintf (tag, sizeof(tag), "darktable|printed|%s", params->prt.printer.name);
   dt_tag_new(tag, &tagid);
-  dt_tag_attach(tagid, imgid);
+  dt_tag_attach(tagid, params->imgid);
+
+  return 0;
+}
+
+static void _print_job_cleanup(void *p)
+{
+  dt_lib_print_job_t *params = p;
+  if(params->pdf_filename[0]) g_unlink(params->pdf_filename);
+  free(params->pdf_image);
+  free(params->pdf_page);
+  free(params->buf);
+  g_free(params->style);
+  g_free(params->buf_icc_profile);
+  g_free(params->p_icc_profile);
+  g_free(params->job_title);
+  free(params);
+}
+
+static void
+_print_button_clicked (GtkWidget *widget, gpointer user_data)
+{
+  const dt_lib_module_t *self = (dt_lib_module_t *)user_data;
+  dt_lib_print_settings_t *ps = (dt_lib_print_settings_t *)self->data;
+
+  const int imgid = dt_view_filmstrip_get_activated_imgid(darktable.view_manager);
+
+  if (imgid == -1)
+  {
+    dt_control_log(_("cannot print until a picture is selected"));
+    return;
+  }
+  if (strlen(ps->prt.printer.name) == 0 || ps->prt.printer.resolution == 0)
+  {
+    dt_control_log(_("cannot print until a printer is selected"));
+    return;
+  }
+  if (ps->prt.paper.width == 0 || ps->prt.paper.height == 0)
+  {
+    dt_control_log(_("cannot print until a paper is selected"));
+    return;
+  }
+
+  dt_job_t *job = dt_control_job_create(&_print_job_run, "print image %d", imgid);
+  if(!job) return;
+
+  dt_lib_print_job_t *params = calloc(1, sizeof(dt_lib_print_job_t));
+  dt_control_job_set_params(job, params, _print_job_cleanup);
+
+  params->imgid = imgid;
+  memcpy(&params->prt, &ps->prt, sizeof(dt_print_info_t));
+
+  // what to call the image?
+  GList *res;
+  if((res = dt_metadata_get(params->imgid, "Xmp.dc.title", NULL)) != NULL)
+  {
+    // FIXME: in metadata_view.c, non-printables are filtered, should we do this here?
+    params->job_title = g_strdup((gchar *)res->data);
+    g_list_free_full(res, &g_free);
+  }
+  else
+  {
+    const dt_image_t *img = dt_image_cache_get(darktable.image_cache, params->imgid, 'r');
+    if(!img)
+    {
+      // in this case no need to release from cache what we couldn't get
+      dt_control_log(_("cannot get image %d for printing"), imgid);
+      dt_control_job_dispose(job);
+      return;
+    }
+    params->job_title = g_strdup(img->filename);
+    dt_image_cache_read_release(darktable.image_cache, img);
+  }
+  // FIXME: ellipsize title/printer as the export completed message is ellipsized
```

Comment:
would be nice to do this before merging.

Reply:
I've done this on my side. Will merge and apply this commit. Thanks.

Reply:
@TurboGit Thank you for fixing, and merging! There are a couple other spots where a run-on title or printer name could be ellipsized:

https://github.com/darktable-org/darktable/blob/019f4b08b7c9d7fb1799405b6e627ad457ef3f90/src/libs/print_settings.c#L241

https://github.com/darktable-org/darktable/blob/019f4b08b7c9d7fb1799405b6e627ad457ef3f90/src/libs/print_settings.c#L436

Do you think these are worth fixing, even via putting a short version of the job title and printer name into `dt_lib_print_job_t`? I'm happy to do that work if you want -- I was away from the computer this weekend hence didn't catch your note until now.

Reply:
Here is a commit which does this work in one place in print, but doesn't provide an overall fix via `dt_control_log`: f093decb8470c994d74d101a729488821c932425

It doesn't properly handle UTF8, though! Maybe there needs to be an overall dt_ellipsize routine, which does the right thing and allocates memory which must be freed later. I like unloading the work to GTK in ffb2d3dcf02259fbdda0525e904a87fbdaf7b4a0, but that would require lots of GUI work for displaying log messages.

Curious what you and houz figure out.

Reply:
@TurboGit wants to add smart ellipsizing of strings that leave the fixed parts alone and only ellipsize the variables that are added via `%s`. I'd assume that for strings that are too big while not containing any variable, or that are not shrunken enough even though the variables are basically replaced by `...` the whole string has to be ellipsized instead.

About offloading the work to GTK, that won't be possible as the message isn't a GTK widget. Instead we can probably use something like this:

https://github.com/darktable-org/darktable/blob/master/src/bauhaus/bauhaus.c#L77-L78

Reply:
This does sound like the right route. Is this something @TurboGit will code? I could look at it, but won't be fast at getting to it.

In the case of print, the message can appear via `dt_control_log` but also in `dt_control_job_add_progress`. The latter doesn't ellipsize long messages but at least crops them, so probably not so big a concern. As long as some of the title shows (which it does) the user can figure out which job is which, in case they want to cancel one.

Reply:
Ah, I see 0a6a32ceebac46b1e74a09a39ab7bc616303ab36 does a better fix for the job labels, too.

Reply:
Another way to handle this, on the model of

https://github.com/darktable-org/darktable/blob/master/src/bauhaus/bauhaus.c#L77-L78

is to just set PANGO_ELLIPSIZE_MIDDLE & width for the whole log message at

https://github.com/darktable-org/darktable/blob/master/src/control/control.c#L263-L264

If the print messages get way to long the user will probably still see the start of the title and the end of the printer name. And this is really an edge case, seems better to handle simply rather than essentially (if I understand the alternative) making a customized printf.

Reply:
I tend to agree.

Reply:
Ok, let's go the simple way then. And we'll see how it works and adjust if necessary.